v-next/hardhat-viem/src/clients.ts

```diff
@@ -0,0 +1,157 @@
+
+export async function getPublicClient<ChainTypeT extends ChainType | string>(
+  provider: EthereumProvider,
+  chainType: ChainTypeT,
+  publicClientConfig?: Partial<viemT.PublicClientConfig>,
+): Promise<GetPublicClientReturnType<ChainTypeT>> {
+  const chain = publicClientConfig?.chain ?? (await getChain(provider));
+  const parameters = {
+    ...getDefaultClientParameters(chain.id),
+    ...publicClientConfig,
+  };
+
+  let publicClient = createPublicClient({
+    chain,
+    transport: customTransport(provider),
+    ...parameters,
+  });
+
+  if (chainType === "optimism") {
+    publicClient = publicClient.extend(publicActionsL2());
+  }
+
+  /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
+  We need to assert the type because TS gets confused with the conditional type */
+  return publicClient as GetPublicClientReturnType<ChainTypeT>;
+}
+
+export async function getWalletClients<ChainTypeT extends ChainType | string>(
+  provider: EthereumProvider,
+  chainType: ChainTypeT,
+  walletClientConfig?: Partial<viemT.WalletClientConfig>,
+): Promise<Array<GetWalletClientReturnType<ChainTypeT>>> {
+  const chain = walletClientConfig?.chain ?? (await getChain(provider));
+  const accounts = await getAccounts(provider);
+  const parameters = {
+    ...getDefaultClientParameters(chain.id),
+    ...walletClientConfig,
+  };
+
+  let walletClients = accounts.map((account) =>
+    createWalletClient({
+      chain,
+      account,
+      transport: customTransport(provider),
+      ...parameters,
+    }),
+  );
+
+  if (chainType === "optimism") {
+    walletClients = walletClients.map((walletClient) =>
+      walletClient.extend(walletActionsL2()),
+    );
+  }
+
+  /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
+  We need to assert the type because TS gets confused with the conditional type */
+  return walletClients as Array<GetWalletClientReturnType<ChainTypeT>>;
+}
+
+export async function getWalletClient<ChainTypeT extends ChainType | string>(
+  provider: EthereumProvider,
+  chainType: ChainTypeT,
+  address: viemT.Address,
+  walletClientConfig?: Partial<viemT.WalletClientConfig>,
+): Promise<GetWalletClientReturnType<ChainTypeT>> {
+  const chain = walletClientConfig?.chain ?? (await getChain(provider));
+  const parameters = {
+    ...getDefaultClientParameters(chain.id),
+    ...walletClientConfig,
+  };
+
+  let walletClient = createWalletClient({
+    chain,
+    account: address,
+    transport: customTransport(provider),
+    ...parameters,
+  });
+
+  if (chainType === "optimism") {
+    walletClient = walletClient.extend(walletActionsL2());
+  }
+
+  /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
+  We need to assert the type because TS gets confused with the conditional type */
+  return walletClient as GetWalletClientReturnType<ChainTypeT>;
+}
+
+export async function getDefaultWalletClient<
```

Comment:
We might want to export this in the future, but for now, it's kept internal in case we decide to take a different approach for retrieving a default account.